File: Python/199.Binary Tree Right Side View.py
#从二叉树的右边看，返回每一层的最右边一个元素
#根据分层返回二叉树元素改编来
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def rightSideView(self, root):
        self.result = []
        self.inorder(root,0)
        return [self.result[i][-1] for i in range(len(self.result))]

    def inorder(self,p,level):
        if p!=None:
            if len(self.result)<=level:
                self.result.append([p.val])
            else:
                self.result[level].append(p.val)
            self.inorder(p.left,level+1) if p.left!=None else None
            self.inorder(p.right,level+1) if p.right!=None else None

t1 = TreeNode(1)
t2 = TreeNode(2)
t3 = TreeNode(3)
t4 = TreeNode(4)
t5 = TreeNode(5)
t6 = TreeNode(6)
t7 = TreeNode(7)
t1.left = t2
t1.right = t3
t2.left = t4
t2.right = t6
t3.right = t5
t3.left = t7

s = Solution()
print(s.rightSideView(t1))

File: C++/Tree/199.Binary Tree Right Side View.cpp
/*
从上到下返回二叉树每一层的最右结点。
*/
/*
思路：修改按层返回结点代码，递归算法。
      但是递归时，先递归右子树，这样只需要在result数组小于level时加入值即可，
      不用进行值的覆盖，由于总是先遍历右边，所以result数组中一定是每层的最右。
*/
#include <iostream>
#include <vector>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
class Solution {
private:
	vector<int> result;
public:
    vector<int> rightSideView(TreeNode* root) {
        helpLevelOrder(root,1);
        return result;
    }
    void helpLevelOrder(TreeNode* cur,int level)
    {
    	if(cur!=NULL)
    	{
    		if(result.size()<level)	result.push_back(cur->val);
    		helpLevelOrder(cur->right,level+1);
    		helpLevelOrder(cur->left,level+1);	
    	}
    }
};
int main()
{
	Solution s;
	TreeNode n1(1),n2(2),n3(3),n4(4),n5(5),n6(6);
	n4.left = &n3, n4.right = &n5;
	n3.left = &n1, n1.right = &n2;
	n5.right = &n6;
	vector<int> out = s.rightSideView(&n4);
	for(int n:out)
		cout<<n<<endl;
}

